File: core/entropass.py
# Copyright (c) 2014 Chris Clark
import re
from base64 import b64encode, b85encode
from hashlib import pbkdf2_hmac


# privateKeyHash is the SHA-512 hash of the private key in lowercase hex
# or the empty string if no private key is set
def generate_password(passphrase, resetCount, privateKeyHash, domain,
                      allowSymbols, length):
    resetCountString = str(resetCount) if resetCount > 0 else ''
    secret = (passphrase + resetCountString + privateKeyHash).encode('utf-8')
    key = pbkdf2_hmac('sha512', secret, domain.encode('utf-8'), 100)
    if allowSymbols:
        return b85encode(key).decode('ascii')[:length]
    else:
        b64 = b64encode(key).decode('ascii')
        return re.sub(r'[\W_]+', '', b64).ljust(len(b64), '0')[:length]

File: webapp/lib/entropass.js
// Copyright (c) 2014 Chris Clark
// requires CryptoJS libraries for SHA-512, PBKDF2, and Base64

var B85CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ' +
               'abcdefghijklmnopqrstuvwxyz' +
               '!#$%&()*+-;<=>?@^_`{|}~';

function b85encodeWord(word) {
    var result = '';
    var powers = [1, 85, 7225, 614125, 52200625];
    for(var i = 0; i < 5; i++)
        result += B85CHARS.charAt(Math.floor((word >>> 0)/powers[4-i]) % 85);
    return result;
}

function b85encode(wordArray) {
    return wordArray.words.map(b85encodeWord).join('');
}

function rpad(str, ch, length) {
    while(str.length < length)
        str += ch;
    return str;
}

function alphanumeric(wordArray) {
    var base64 = CryptoJS.enc.Base64.stringify(wordArray);
    return rpad(base64.replace(/[_\W]/g, ''), '0', base64.length);
}

// privateKeyHash is the SHA-512 hash of the private key in lowercase hex
// or the empty string if no private key is set
function generatePassword(passphrase, resetCount, privateKeyHash, domain,
                           allowSymbols, length) {
    var resetCountString = resetCount > 0 ? resetCount.toString() : '';
    var secret = passphrase + resetCountString + (privateKeyHash || '');
    var hash = CryptoJS.PBKDF2(secret, domain, {keySize: 512/32,
                               iterations: 100, hasher: CryptoJS.algo.SHA512});
    var password = allowSymbols ? b85encode(hash) : alphanumeric(hash);
    return password.slice(0, length);
}
